spark/src/main/java/org/apache/iceberg/spark/data/SparkOrcReader.java

```diff
@@ -53,425 +40,83 @@
  * It minimizes allocations by reusing most of the objects in the implementation.
  */
 public class SparkOrcReader implements OrcValueReader<InternalRow> {
-  private static final int INITIAL_SIZE = 128 * 1024;
-  private final List<TypeDescription> columns;
-  private final Converter[] converters;
-  private final UnsafeRowWriter rowWriter;
+  private final SparkOrcValueReaders.StructReader reader;
 
-  public SparkOrcReader(TypeDescription readOrcSchema) {
-    columns = readOrcSchema.getChildren();
-    converters = buildConverters();
-    rowWriter = new UnsafeRowWriter(columns.size(), INITIAL_SIZE);
+  public SparkOrcReader(org.apache.iceberg.Schema expectedSchema, TypeDescription readSchema) {
+    this(expectedSchema, readSchema, ImmutableMap.of());
   }
 
-  private Converter[] buildConverters() {
-    final Converter[] newConverters = new Converter[columns.size()];
-    for (int c = 0; c < newConverters.length; ++c) {
-      newConverters[c] = buildConverter(columns.get(c));
-    }
-    return newConverters;
+  @SuppressWarnings("unchecked")
+  public SparkOrcReader(
+      org.apache.iceberg.Schema expectedSchema, TypeDescription readOrcSchema, Map<Integer, ?> idToConstant) {
+    reader = (SparkOrcValueReaders.StructReader) OrcSchemaWithTypeVisitor.visit(
+        expectedSchema, readOrcSchema, new ReadBuilder(idToConstant));
   }
 
   @Override
   public InternalRow read(VectorizedRowBatch batch, int row) {
-    rowWriter.reset();
-    rowWriter.zeroOutNullBytes();
-    for (int c = 0; c < batch.cols.length; ++c) {
-      converters[c].convert(rowWriter, c, batch.cols[c], row);
-    }
-    return rowWriter.getRow();
+    return reader.read(batch, row);
```

Comment:
This is where you could wrap the columns from `VectorizedRowBatch` in a `StructColumnVector`.

Reply:
It worked!